File: output/sonarqube_analysis/cursor_solutions/python/results/q37-longest-valid-parentheses.py
# https://leetcode.com/problems/longest-valid-parentheses/
# Problem: Longest Valid Parentheses

#Problem Description:
#Given a string containing just the characters '(' and ')', return the length of the longest valid (well-formed) parentheses substring.

# Constraints:
# 0 <= s.length <= 3 * 10^4
# s[i] is '(', or ')'.

#Code Structure:
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s:
            return 0
        
        stack = [-1]  # Initialize with -1 to handle edge cases
        max_length = 0
        
        for i, char in enumerate(s):
            if char == '(':
                stack.append(i)
            else:  # char == ')'
                stack.pop()
                if not stack:
                    # No matching opening bracket, update the base
                    stack.append(i)
                else:
                    # Calculate the length of valid parentheses
                    max_length = max(max_length, i - stack[-1])
        
        return max_length
    

File: output/sonarqube_analysis/cursor_solutions/php/results/q37-longest-valid-parentheses.php
<?php
/** https://leetcode.com/problems/longest-valid-parentheses */
// Problem: Longest Valid Parentheses

//Problem Description:
//Given a string containing just the characters '(' and ')', return the length of the longest valid (well-formed) parentheses substring.

// Constraints:
// 0 <= s.length <= 3 * 10^4
// s[i] is '(', or ')'.


//Code Structure:
class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function longestValidParentheses($s) {
        $n = strlen($s);
        if ($n <= 1) {
            return 0;
        }
        
        $dp = array_fill(0, $n, 0);
        $maxLen = 0;
        
        for ($i = 1; $i < $n; $i++) {
            if ($s[$i] === ')') {
                if ($s[$i - 1] === '(') {
                    // Case: ...()
                    $dp[$i] = ($i >= 2 ? $dp[$i - 2] : 0) + 2;
                } else if ($i - $dp[$i - 1] > 0 && $s[$i - $dp[$i - 1] - 1] === '(') {
                    // Case: ...((...))
                    $dp[$i] = $dp[$i - 1] + (($i - $dp[$i - 1] >= 2) ? $dp[$i - $dp[$i - 1] - 2] : 0) + 2;
                }
                $maxLen = max($maxLen, $dp[$i]);
            }
        }
        
        return $maxLen;
    }
}
    